sdk-extensions/incubator/src/test/java/io/opentelemetry/sdk/extension/incubator/fileconfig/ConfigurationReaderTest.java

```diff
@@ -337,4 +344,149 @@ void nullValuesParsedToEmptyObjects() {
+            + "      ratio:\n"; // Double
+
+    OpenTelemetryConfiguration model =
+        ConfigurationReader.parse(new ByteArrayInputStream(yaml.getBytes(StandardCharsets.UTF_8)));
+
+    assertThat(model.getFileFormat()).isNull();
+    assertThat(model.getDisabled()).isNull();
+    assertThat(model.getAttributeLimits().getAttributeValueLengthLimit()).isNull();
+    assertThat(model.getTracerProvider().getSampler().getTraceIdRatioBased().getRatio()).isNull();
+
+    assertThat(model)
+        .isEqualTo(
+            new OpenTelemetryConfiguration()
+                .withAttributeLimits(new AttributeLimits())
+                .withTracerProvider(
+                    new TracerProvider()
+                        .withSampler(
+                            new Sampler().withTraceIdRatioBased(new TraceIdRatioBased()))));
+  }
+
+  @ParameterizedTest
+  @MethodSource("envVarSubstitutionArgs")
+  void envSubstituteAndLoadYaml(
+      String rawYaml, String expectedSubstituteResult, Object expectedYamlResult) {
+    Map<String, String> environmentVariables = new HashMap<>();
+    environmentVariables.put("VAR_1", "value1");
+    environmentVariables.put("VAR_2", "value2");
+
+    String substitutedYaml =
+        ConfigurationReader.substituteEnvVariables(
+            new ByteArrayInputStream(rawYaml.getBytes(StandardCharsets.UTF_8)),
+            environmentVariables);
+    assertThat(substitutedYaml).isEqualTo(expectedSubstituteResult);
+
+    Object yaml =
+        ConfigurationReader.loadYaml(
+            new ByteArrayInputStream(rawYaml.getBytes(StandardCharsets.UTF_8)),
+            environmentVariables);
+    assertThat(yaml).isEqualTo(expectedYamlResult);
+  }
+
+  @SuppressWarnings("unchecked")
+  private static java.util.stream.Stream<Arguments> envVarSubstitutionArgs() {
+    return java.util.stream.Stream.of(
+        // Simple cases
+        Arguments.of("key1: ${env:VAR_1}", "key1: value1\n", mapOf(entry("key1", "value1"))),
+        Arguments.of(
+            "key1: ${env:VAR_1}\nkey2: value2\n",
+            "key1: value1\nkey2: value2\n",
+            mapOf(entry("key1", "value1"), entry("key2", "value2"))),
+        Arguments.of(
+            "key1: ${env:VAR_1} value1\nkey2: value2\n",
+            "key1: value1 value1\nkey2: value2\n",
+            mapOf(entry("key1", "value1 value1"), entry("key2", "value2"))),
+        // Multiple environment variables referenced
+        Arguments.of(
+            "key1: ${env:VAR_1}${env:VAR_2}\nkey2: value2\n",
+            "key1: value1value2\nkey2: value2\n",
+            mapOf(entry("key1", "value1value2"), entry("key2", "value2"))),
+        Arguments.of(
+            "key1: ${env:VAR_1} ${env:VAR_2}\nkey2: value2\n",
+            "key1: value1 value2\nkey2: value2\n",
+            mapOf(entry("key1", "value1 value2"), entry("key2", "value2"))),
+        // VAR_3 is not defined in environment
+        Arguments.of(
+            "key1: ${env:VAR_3}\nkey2: value2\n",
```

Comment:
These multi-line yamls that have been collapsed into a single one-line string with `\n` in the middle are a bit hard to read. I think a string concat would read slightly better here.